upload-api/tables/delegations.js

```diff
@@ -134,27 +151,54 @@ async function cidToDelegation (bucket, cid) {
   return delegation
 }
 
-/**
- * 
- * @param {import('../types').DelegationsBucket} bucket
- * @param {Ucanto.Delegation<Ucanto.Tuple<Ucanto.Capability>>[]} delegations
+/** 
+ * @param {import('../types').InvocationBucket} invocationBucket
+ * @param {import('../types').WorkflowBucket} workflowBucket
+ * @param {CID} invocationCid
+ * @param {CID} delegationCid
+ * @returns {Promise<Ucanto.Delegation>}
  */
-async function writeDelegations (bucket, delegations) {
-  return writeEntries(
-    bucket,
-    [...delegations].map((delegation) => {
-      const carBytes = delegationsToBytes([delegation])
-      const value = carBytes
-      return /** @type {[key: CID, value: Uint8Array]} */ ([delegation.cid, value])
-    })
+async function findDelegationInInvocation (invocationBucket, workflowBucket, invocationCid, delegationCid) {
+
+  // TODO is this right? I cargo culted some code from ucan-invocation.js and w3up/upload-api/src/access/delegate.js
+  // but am not highly confident it works
+
+  const invocationCidStr = invocationCid.toString()
+  const agentMessageWithInvocationCid = await invocationBucket.getInLink(
+    invocationCid.toString()
   )
-}
 
-/**
- * 
- * @param {import('../types').DelegationsBucket} bucket
- * @param {Iterable<readonly [key: CID, value: Uint8Array ]>} entries
- */
-async function writeEntries (bucket, entries) {
-  await Promise.all([...entries].map(([key, value]) => bucket.put(key, value)))
+  if (!agentMessageWithInvocationCid) {
+    throw new NoInvocationFoundForGivenCidError()
+  }
+
+  const agentMessageBytes = await workflowBucket.get(
+    agentMessageWithInvocationCid
+  )
+  if (!agentMessageBytes) {
+    throw new NoInvocationFoundForGivenCidError()
+  }
+
+  const agentMessage = await CAR.request.decode({
```

Comment:
This works, however not exactly how I'd expect one to use it. I think doing [`CAR.decode`](https://github.com/web3-storage/ucanto/blob/d6f1f7b8e8d97db8515a8b32d4b687102580418d/packages/core/src/car.js#L97-L119) first and then [`Message.view`](https://github.com/web3-storage/ucanto/blob/main/packages/core/src/message.js#L32-L40) would be expected use of APIs (where `store` is `blocks` returned by `CAR.decode`).

That is more or less what [`CAR.request.decode` does](https://github.com/web3-storage/ucanto/blob/main/packages/transport/src/car/request.js#L45-L56), although it may change in the future, which is why I don't think it should be used here.

Reply:
Taking closer look this seems unnecessarily complicated way to go about this. It could be simplified by:

1. Doing `const { blocks } = CAR.decode(bytes)` (where CAR is from `core` as opposed to transport).
2. Calling `const delegation = Delegation.view({ blocks, root: delegationCID })`

It is also going to be completely immune to `AgentMessage` format changes, because it simply creates a cursor into CAR blocks.

⚠️ Also once you create a delegation in 2nd step I would also access `delegation.data` to force delegation to be materialized. This is unfortunate side effect of view been lazy, which I'll address someday.

Reply:
ok great - chatted in a huddle and got this looking good - much simpler, thanks!!